ebu_tt_live/documents/ebutt3.py

```diff
@@ -81,9 +178,100 @@ def availability_time(self, value):
+            sequence_number=self.sequence_number)
+        )
+        self._resolved_end_time = resolved_end_time
+        self._resolved_begin_time = self.computed_begin_time
+
+    @property
+    def resolved_end_time(self):
+        if self._resolved_end_time is not None:
+            return self._resolved_end_time
+        else:
+            return self.sequence.resolved_end_time(self)
+
+    @property
+    def time_base(self):
+        return self._ebutt3_content.timeBase
+
+    @property
+    def discarded(self):
+        return self.resolved_begin_time >= self.resolved_end_time
 
     def validate(self):
-        self._ebutt3_content.validateBinding()
+        # This is assuming availability from the beginning of our time coordinate system.
+        availability_time = self.availability_time or timedelta()
+        # Run validation
+        result = self._ebutt3_content.validateBinding(
+            availability_time=availability_time
+        )
+        # Extract results
+
+        # Begin times
+        bfs_queue = []
+        # Default value for the computed begin time of the document is the active begin time of the body
+        # This only changes if the body does not declare a begin time.
+        self._computed_begin_time = self._ebutt3_content.body.computed_begin_time
+        body_begin = self._ebutt3_content.body.begin
+        # If the body defines no begin time, the document's resolved begin time
+        # is the earliest begin time of body's children elements (and this
+        # applies recursively)
+        if body_begin is None:
+            children = self._ebutt3_content.body.orderedContent()
+            for child in children:
+                bfs_queue.append(child.value)
+            min_child_begin = None
+            # BFS over the tree of elements, starting at body
+            while bfs_queue:
+                current_child = bfs_queue.pop(0)
+                if isinstance(current_child, TimeBaseValidationMixin):
+                    # if current child declares a begin time
+                    if current_child.begin is not None:
+                        # if it is the first child we see to declare a begin
+                        # time. We take it as our current earlier computed
+                        # begin time. Otherwise we compare with the one we
+                        # have.
+                        if min_child_begin is None:
+                            min_child_begin = current_child.computed_begin_time
+                        else:
+                            if current_child.computed_begin_time < min_child_begin:
+                                min_child_begin = current_child.computed_begin_time
+                    current_child_children = current_child.orderedContent()
+                    for child in current_child_children:
+                        bfs_queue.append(child.value)
+            if min_child_begin is not None:
+                self._computed_begin_time = min_child_begin
```

Comment:
@kozmaz87 all this block is the way I handled computation of the computed begin times of the document. I handle it here because after the discussion with Nigel I realized that the document's computed begin time can be different from the body's active begin time. So now the document's computed begin time is the active begin time of the first element declaring a `begin` attribute if there is one such element.